src/vm/amd64/asmhelpers.S

```diff
@@ -0,0 +1,195 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+.intel_syntax noprefix
+#include "unixasmmacros.inc"
+#include "asmconstants.h"
+
+#define real4 dword
+#define real8 qword
+
+//
+//    typedef struct _PROFILE_PLATFORM_SPECIFIC_DATA
+//    {
+//        FunctionID *functionId; // function ID comes in the r11 register
+//        void       *rbp;
+//        void       *probersp;
+//        void       *ip;
+//        void       *profiledRsp;
+//        UINT64      rax;
+//        LPVOID      hiddenArg;
+//        UINT64      flt0;
+//        UINT64      flt1;
+//        UINT64      flt2;
+//        UINT64      flt3;
+//        UINT32      flags;
+//    } PROFILE_PLATFORM_SPECIFIC_DATA, *PPROFILE_PLATFORM_SPECIFIC_DATA;
+//
+.equ SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA, 0x8*11 + 0x4*2    // includes fudge to make FP_SPILL right
+.equ SIZEOF_OUTGOING_ARGUMENT_HOMES, 0x8*6
+.equ SIZEOF_FP_ARG_SPILL, 0x10*1
+
+// Need to be careful to keep the stack 16byte aligned here, since we are pushing 3
+// arguments that will align the stack and we just want to keep it aligned with our
+// SIZEOF_STACK_FRAME
+
+.equ OFFSETOF_PLATFORM_SPECIFIC_DATA, SIZEOF_OUTGOING_ARGUMENT_HOMES
+
+// we'll just spill into the PROFILE_PLATFORM_SPECIFIC_DATA structure
+.equ OFFSETOF_FP_ARG_SPILL, SIZEOF_OUTGOING_ARGUMENT_HOMES + SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA
+
+.equ SIZEOF_STACK_FRAME, SIZEOF_OUTGOING_ARGUMENT_HOMES + SIZEOF_PROFILE_PLATFORM_SPECIFIC_DATA + SIZEOF_MAX_FP_ARG_SPILL
+
+.equ PROFILE_ENTER, 0x1
+.equ PROFILE_LEAVE, 0x2
+.equ PROFILE_TAILCALL, 0x4
+
+// ***********************************************************
+//   NOTE:
+//
+//   Register preservation scheme:
+//
+//       Preserved:
+//           - all non-volatile registers
+//           - rax
+//           - xmm0
+//
+//       Not Preserved:
+//           - integer argument registers (rcx, rdx, r8, r9)
+//           - floating point argument registers (xmm1-3)
+//           - volatile integer registers (r10, r11)
+//           - volatile floating point registers (xmm4-5)
+//
+// ***********************************************************
+
+// EXTERN_C void ProfileEnterNaked(FunctionIDOrClientID functionIDOrClientID, size_t profiledRsp);
+// <NOTE>
+//
+// </NOTE>
+NESTED_ENTRY ProfileEnterNaked, _TEXT, NoHandler
+  //       Upon entry :
+  //           rdi = clientInfo
+  //           rsi = profiledRsp
+
+  PUSH_CALLEE_SAVED_REGISTERS
+  push_argument_register  r9
+  push_argument_register  r8
+  push_argument_register  rcx
+  push_argument_register  rdx
+  push_nonvol_reg         r10
+  push_nonvol_reg         rax
+
+  lea                     rax, [rsp + 0x10]    // caller rsp
```

Comment:
I rarely deal with assembly, but it doesn't seem like this would calculate the caller's rsp correctly?

At this point in the function we have already pushed r15, r14, r13, r12, rbp, rbx, rsi, rdi, r10, and rax. Did those pushes generate code that doesn't update rsp?